compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -330,6 +344,24 @@ fn emit_orphan_check_error<'tcx>(
     })
 }
 
+fn emit_orphan_check_for_impl_on_types<'tcx, T: rustc_errors::EmissionGuarantee>(
+    full_impl_span: Span,
+    self_ty_span: Span,
+    ptr_ty: &ty::TypeAndMut<'_>,
+    diag: &mut DiagnosticBuilder<'tcx, T>,
+) {
+    let mut_key = if ptr_ty.mutbl == rustc_middle::mir::Mutability::Mut { "mut " } else { "" };
+    let msg_sugg = "consider to use a new wrapper type".to_owned();
+    let sugg = vec![
+        (
+            full_impl_span.shrink_to_lo(),
+            format!("struct WrapperType(*{}{});\n\n", mut_key, ptr_ty.ty),
+        ),
+        (self_ty_span, "WrapperType".to_owned()),
```

Comment:
This does not work if the self type has generics:

```rust
impl<T> ForeignTrait for *mut T {}
```

Reply:
Lets try to skip this case for this case, and put this also inside the test UI to make sure that this will be cover